### PR TITLE
FEA parallel_config context manager to allow more fine-grained control

### PR DESCRIPTION
**Pull request by jeremiedbb**
Fixes https://github.com/joblib/joblib/issues/912
Fixes https://github.com/joblib/joblib/issues/979
related to https://github.com/joblib/joblib/issues/1376

adds a new context_manager `parallel_config` that should be used in place of `parallel_backend`. It exposes more parameters and does not enforce setting a backend.

It allows to do something like:
```py
from joblib.parallel import parallel_config
from joblib.memory import Memory
from joblib import Parallel, delayed
 
with parallel_config(prefer="threads", n_jobs=3):
    res = Parallel(verbose=100)(delayed(lambda :1)() for _ in range(10))
 
# [Parallel(n_jobs=3)]: Using backend ThreadingBackend with 3 concurrent workers.
# [Parallel(n_jobs=3)]: Done   1 tasks      | elapsed:    0.0s
# [Parallel(n_jobs=3)]: Done   2 tasks      | elapsed:    0.0s
# [Parallel(n_jobs=3)]: Done   3 tasks      | elapsed:    0.0s
# [Parallel(n_jobs=3)]: Done   4 tasks      | elapsed:    0.0s
# [Parallel(n_jobs=3)]: Done   5 tasks      | elapsed:    0.0s
# [Parallel(n_jobs=3)]: Done   6 out of  10 | elapsed:    0.0s remaining:    0.0s
# [Parallel(n_jobs=3)]: Done   7 out of  10 | elapsed:    0.0s remaining:    0.0s
# [Parallel(n_jobs=3)]: Done   8 out of  10 | elapsed:    0.0s remaining:    0.0s
# [Parallel(n_jobs=3)]: Done  10 out of  10 | elapsed:    0.0s remaining:    0.0s
# [Parallel(n_jobs=3)]: Done  10 out of  10 | elapsed:    0.0s finished
```
It's using the threading backend instead of the default loky, with 3 jobs instead of the default 1.

The config parameters of the context are used only if they are not explicitely set in `Parallel`. It makes us forced to use sentinels for the default values, adding a little bit of complexity.

TODO
- [x] raise an error if backend constructor params are set but backend is left to None
- [x] add tests

### CODE REVIEW

**Review by tomMoral — COMMENTED** [review]
From a quick scan, the setup looks nice :)
Just a few first comments.

File: joblib/parallel.py
Comment (jeremiedbb) [reply joblib/parallel.py:812]: haha don't worry we'll blackify all joblib at 6pm :)

File: joblib/parallel.py
Comment (jeremiedbb) [reply joblib/parallel.py:212]: It looks like this
```py
signature(Parallel)
<Signature (n_jobs=default(None), backend=None, verbose=default(0), timeout=None, pre_dispatch='2 * n_jobs', batch_size='auto', temp_folder=default(None), max_nbytes=default('1M'), mmap_mode=default('r'), prefer=default(None), require=default(None))>
```

Comment (jeremiedbb) [reply joblib/parallel.py:212]: > I guess this is what is taken care of with Sentinel.__repr__?

absolutely :)

File: joblib/parallel.py
Comment (jjerphan) [inline joblib/parallel.py:862]: In this PR, the introduction of default values of back-end parameters comes with some duplication. Moreover, `backend_params` and `default_parallel_config` (the dict wrapping the configuration's parameters' values) are also passed or unpacked at several places. 

Hence a naive question: would using instances small `dataclass` (e.g. `BackendParams`) which would wrap the mapping and potentially the logic (that is being duplicated here and in other places) help?

File: joblib/parallel.py
Comment (jjerphan) [inline joblib/parallel.py:363]: As you pointed out in a discussion, this can be removed?

File: joblib/parallel.py
Comment (jjerphan) [inline joblib/parallel.py:124]: What is the reason for having a public function and a private function, here?

**Review by jjerphan — COMMENTED** [review]
To me, this is a valuable contribution for users. Thank you, @jeremiedbb.

Here is a review from an outsider (I am starting getting around joblib).

File: joblib/parallel.py
Comment (jeremiedbb) [reply joblib/parallel.py:363]: already did :)

File: joblib/parallel.py
Comment (jeremiedbb) [reply joblib/parallel.py:983]: I'm not sure what you mean. Can you give an example of how it would look like ?

Comment (jjerphan) [reply joblib/parallel.py:983]: Nevermind: what I proposed above would make the code more complex for nothing. You can resolve this thread.

File: joblib/parallel.py
Comment (glemaitre) [inline joblib/parallel.py:117]: I did not write it for the moment, but we could eventually come up with a more concise way to write this function using dictionary updates keeping the priority rules.

File: joblib/parallel.py
Comment (glemaitre) [inline joblib/parallel.py:103]: A small docstring could help here.

File: joblib/parallel.py
Comment (glemaitre) [inline joblib/parallel.py:121]: Currently, `_Sentinel` is a mutable object and we used it as a default parameter. It would be safer to make sure this is immutable. We can probably achieve this goal in several ways: using `__slots__` or decorating with the `dataclass` decorator (`frozen` could be useful there - maybe `slots` as well).

File: joblib/parallel.py
Comment (glemaitre) [inline joblib/parallel.py:125]: Could be nice to have a fully documented docstring for the public function.

File: joblib/parallel.py
Comment (glemaitre) [inline joblib/parallel.py:200]:
```suggestion
    """Change the default backend or configuration used by :class:`~Parallel`.
```

File: joblib/parallel.py
Comment (glemaitre) [inline joblib/parallel.py:211]: Let's put this into an `Examples` section.

File: joblib/parallel.py
Comment (glemaitre) [inline joblib/parallel.py:247]: I am wondering if it could be another section in the `Examples` section

File: joblib/parallel.py
Comment (glemaitre) [inline joblib/parallel.py:259]:
```suggestion
        The maximum number of concurrently running jobs, such as the number
        of Python worker processes when `backend="multiprocessing"`
        or the size of the thread-pool when `backend="threading"`.
        If -1 all CPUs are used. If 1 is given, no parallel computing code
        is used at all, which is useful for debugging. For `n_jobs` below -1,
        (n_cpus + 1 + n_jobs) are used. Thus for `n_jobs=-2`, all
        CPUs but one are used.
        `None` is a marker for 'unset' that will be interpreted as `n_jobs=1`.
```

File: joblib/parallel.py
Comment (glemaitre) [inline joblib/parallel.py:279]:
```suggestion
        Folder to be used by the pool for memmapping large arrays
        for sharing memory with worker processes. If None, this will try in
        order:

        - a folder pointed by the `JOBLIB_TEMP_FOLDER` environment
          variable,
        - `/dev/shm` if the folder exists and is writable: this is a
          RAM disk filesystem available by default on modern Linux
          distributions,
        - the default system temporary folder that can be
          overridden with `TMP`, `TMPDIR` or `TEMP` environment
          variables, typically `/tmp` under Unix operating systems.
```

File: joblib/parallel.py
Comment (glemaitre) [inline joblib/parallel.py:290]: It could be nice to make this link clickable.

File: joblib/parallel.py
Comment (glemaitre) [inline joblib/parallel.py:303]:
```suggestion
        `parallel_backend`.
```

File: joblib/parallel.py
Comment (glemaitre) [inline joblib/parallel.py:323]: I think it should be 1.3. It would be better to add it after the long summary above.
```suggestion

    .. versionadded:: 1.3
```

File: joblib/parallel.py
Comment (glemaitre) [inline joblib/parallel.py:382]:
```suggestion
                    raise ValueError(
                        "Invalid backend: {backend}, expected one of {sorted(BACKENDS.keys()):!r}"
                    )
```

File: joblib/parallel.py
Comment (glemaitre) [inline joblib/parallel.py:388]:
```suggestion
            msg = (
                f"{backend.__class__.__name__} does not accept setting the "
                "inner_max_num_threads argument"
            )
```

File: joblib/parallel.py
Comment (glemaitre) [inline joblib/parallel.py:389]: Do we want to raise a `ValueError` instead of an `AssertionError` here?

File: joblib/parallel.py
Comment (glemaitre) [inline joblib/parallel.py:423]: Maybe a `See Also` section would be best to discover it?

File: joblib/parallel.py
Comment (glemaitre) [inline joblib/parallel.py:983]: I think we can make a for loop here and directly create a dictionary because this is already something that we do below. So a dict comprehension seems possible here.

File: joblib/parallel.py
Comment (glemaitre) [inline joblib/parallel.py:313]: We could also have a `See Also` to point to `parallel_backend`.

File: joblib/parallel.py
Comment (jeremiedbb) [reply joblib/parallel.py:125]: I left the docstring as is. Maybe a task for a separate PR ?

File: joblib/parallel.py
Comment (jeremiedbb) [reply joblib/parallel.py:290]: it is on joblib's doc, see https://joblib.readthedocs.io/en/latest/generated/joblib.Parallel.html

File: joblib/parallel.py
Comment (jeremiedbb) [reply joblib/parallel.py:389]: I didn't want to change the current behavior

File: joblib/parallel.py
Comment (jeremiedbb) [reply joblib/parallel.py:313]: It's good to have a See Also from parallel_backend to parallel_config but I don't think we want the other way around since we say that parallel_config should be used in place of parallel_backend

File: joblib/_utils.py
Comment (jeremiedbb) [reply joblib/_utils.py]: I changed it to use a dataclass

File: joblib/parallel.py
Comment (jeremiedbb) [reply joblib/parallel.py:126]: used a frozen dataclass

File: joblib/parallel.py
Comment (ogrisel) [inline joblib/parallel.py:210]:
```suggestion
    This is an alternative to directly passing keyword arguments to the
    :class:`~Parallel` class constructor. It is particularly useful when calling
    into library code that uses joblib internally but does not expose the various
    parallel configuration arguments in its own API/
```

File: joblib/parallel.py
Comment (ogrisel) [inline joblib/parallel.py:299]:
```suggestion
        MKL or OpenMP. This is only used with the `loky` backend.
```

**Comment by ogrisel** [comment]
@jeremiedbb There are a few conflicts to resolve after the merge of the `return_generator` branch but it does not look too complicated.

File: joblib/parallel.py
Comment (tomMoral) [reply joblib/parallel.py:353]: Amazing!

File: joblib/parallel.py
Comment (tomMoral) [inline joblib/parallel.py:393]:
```suggestion
                        stacklevel=2
                    )
```

File: joblib/parallel.py
Comment (tomMoral) [inline joblib/parallel.py:1191]: can't we use the `_get_config_param` here because of the `None` value?
Not sure any one would use `None` so I would use `_get_config_param` no?

**Review by tomMoral — APPROVED** [review]
I did a pass but the code looks super nice!! Good job @jeremiedbb :)

File: joblib/parallel.py
Comment (jeremiedbb) [reply joblib/parallel.py]: I added a comment, but we can't merge the 2 conditions because there's some code running regardless of the second condition

File: joblib/parallel.py
Comment (jeremiedbb) [reply joblib/parallel.py:291]: I didn't want to break backward compat. It's not really related to this PR so I think it should be considered in a separate issue/PR

File: joblib/parallel.py
Comment (jeremiedbb) [reply joblib/parallel.py:94]: I agree it's not ideal, but there's a fair amount of entanglement between the different objects in this file and it's hard to split them without creating cyclic imports. It would at least require a third file containing the common things. Maybe better to leave that for another PR ?

Comment (tomMoral) [reply joblib/parallel.py:94]: agreed.

File: joblib/parallel.py
Comment (tomMoral) [reply joblib/parallel.py:291]: Yes alright, we need to list all these tasks in an issue maybe as we have several comment saying: for the next PR :)

**Review by tomMoral — APPROVED** [review]
LGTM! thanks @jeremiedbb for making this :)

File: joblib/parallel.py
Comment (tomMoral) [reply joblib/parallel.py]: Indeed, sorry I misread the indentation ^^' and great to add the comment.

**Comment by trendelkampschroer** [comment]
Thanks a lot for this feature.

**Pull request merged**
